s2spy/traintest.py

```diff
@@ -2,52 +2,94 @@
 
 A collection of train/test splitting approaches for cross-validation.
 """
+from typing import Union
+import numpy as np
 import pandas as pd
-from sklearn.model_selection import KFold
+import xarray as xr
 
 
-def kfold(df: pd.DataFrame, n_splits: int = 2, shuffle=False, **kwargs):
-    """
-    K-Folds cross-validator, which splits provided intervals into k
-    consecutive folds.
-
-    For more information
+def fold_by_anchor(cv, data: Union[xr.Dataset, pd.DataFrame]):
+    """Splits calendar resampled data into train/test groups, based on the anchor year.
+    As splitter, a Splitter Class such as sklearn's KFold can be passed.
 
-    args:
-        df: DataFrame to store the train/test groups. It must have a column named "anchor_years".
-        n_splits: number of train/test splits.
+    For an overview of the sklearn Splitter Classes see:
+    https://scikit-learn.org/stable/modules/classes.html#module-sklearn.model_selection
 
-    Other keyword arguments: see the documentation for sklearn.model_selection.KFold:
-    https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html
+    Args:
+        cv (Splitter Class): Initialized splitter class, much have a `fit(X)` method
+            which splits up `X` into multiple folds of train/test data.
+        data (xr.Dataset or pd.DataFrame): Dataset with `anchor_year` as dimension or
+            DataFrame with `anchor_year` as column. E.g. data resampled using
+            `s2spy.time.AdventCalendar`'s `resample` method.
 
     Returns:
-        Pandas DataFrame with columns containing train/test label in each fold.
+        The input dataset with extra coordinates or columns added for each fold,
+            containing the labels 'train', 'test', and possibly 'skip'.
     """
-    cv = KFold(n_splits, shuffle=shuffle, **kwargs)
-    folds = cv.split(df)
+    assert isinstance(
+        data, (xr.Dataset, pd.DataFrame)
+    ), "Input data should be of type xr.Dataset or pd.DataFrame"
 
-    # Map folds to a new dataframe
-    output = pd.DataFrame(index=df.index)
-    for i, (train_index, test_index) in enumerate(folds):
-        col_name = f"fold_{i}"
-        output[col_name] = "skip"
-        output[col_name].iloc[train_index] = "train"
-        output[col_name].iloc[test_index] = "test"
+    if isinstance(data, xr.Dataset):
+        data = data.copy()  # otherwise the dataset is modified inplace
 
-    return output
+        assert (
```

Comment:
I'm not sure where I got it from, but I always thought it bad practice to use assert statements outside of tests. See e.g. here for some relevant discussion: https://community.sonarsource.com/t/feature-python-assert-should-be-consider-harmful/38501/3.

Some thoughts:
+ Assertions can be short and simple
- Assertions can be removed in optimized production code
- Assertions always raise assertionerrors as opposed to a more befitting error message.

An alternative and equally short solution would be:

```py
if len(data.anchor_year.size) < 1:
    raise InputError("Input data must have more than 1 anchor year to split")
```

with the only upside/downside that you can/need to define the `InputError` (or you can be lazy and use a ValueError instead)